src/Plugin/Block/CorporateFormBlock.php

```diff
@@ -0,0 +1,147 @@
+use Drupal\contact\Entity\ContactForm;
+use Drupal\Core\Entity\EntityFormBuilderInterface;
+
+/**
+ * Provides a block that renders a corporate form.
+ *
+ * @Block(
+ *   id = "oe_contact_forms_corporate_block",
+ *   admin_label = @Translation("Corporate form"),
+ *   category = @Translation("Corporate forms"),
+ *   deriver = "Drupal\oe_contact_forms\Plugin\Derivative\CorporateFormBlock",
+ * )
+ */
+class CorporateFormBlock extends BlockBase implements ContainerFactoryPluginInterface {
+
+  /**
+   * The EntityFormBuilder.
+   *
+   * @var \Drupal\Core\Entity\EntityFormBuilder
+   */
+  protected $entityFormBuilder;
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * Constructs a new CorporateFormBlock.
+   *
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin_id for the plugin instance.
+   * @param array $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entityTypeManager
+   *   The entity type manager.
+   * @param \Drupal\Core\Entity\EntityFormBuilderInterface $entity_form_builder
+   *   The entity form builder.
+   */
+  public function __construct(array $configuration, string $plugin_id, array $plugin_definition, EntityTypeManagerInterface $entityTypeManager, EntityFormBuilderInterface $entity_form_builder) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->entityTypeManager = $entityTypeManager;
+    $this->entityFormBuilder = $entity_form_builder;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('entity_type.manager'),
+      $container->get('entity.form_builder')
+    );
+  }
+
+  /**
+   * When the contact form is modified, these cache tags will be invalidated.
+   */
+  public function getCacheTags() {
+    /** @var \Drupal\contact\ContactFormInterface $contact_form */
+    $contact_form = $this->getContactForm();
+    if (!$contact_form) {
+      return [];
+    }
+
+    return Cache::mergeTags(parent::getCacheTags(), $contact_form->getCacheTags());
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function blockAccess(AccountInterface $account) {
+    /** @var \Drupal\contact\ContactFormInterface $contact_form */
+    $contact_form = $this->getContactForm();
+
+    if (!$contact_form) {
+      return AccessResult::forbidden();
+    }
+
+    $access = $this->entityTypeManager->getAccessControlHandler('contact_message')->createAccess($contact_form->id(), $account);
+
+    if (!$access) {
+      return AccessResult::forbidden()->cachePerPermissions();
```

Comment:
Return directly the access object as it may contain other cache metadata. Same for the allowed below. Just add to the existing cache metadata.

Reply:
so I return AccessResult or $access? not clear for me here

Reply:
$access already contains cache metadata, so return that of course + add onto it your own cache dependencies if needed.

Reply:
done, tested, nothing additional is needded